File: 0046-permutations/0046-permutations.cpp
class Solution {
private: 
    void permute(vector<int>& nums, int index, vector<vector<int>>& res)
    {
        if(index == nums.size())
        {
            res.push_back(nums);
            return;
        }
        
        for(int i=index; i< nums.size(); i++)
        {
            swap(nums[index], nums[i]);
            permute(nums, index + 1, res);
            swap(nums[i], nums[index]);
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        
        vector<vector<int>>res;
        permute(nums, 0, res);
        return res;
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        res = []
        def permute(temp, index):
                    
            if index == len(temp):
                res.append(temp.copy())
                
            for i in range(index, len(temp)): 
                temp[index], temp[i] = temp[i], temp[index]
                permute(temp, index + 1)
                temp[index], temp[i] = temp[i], temp[index]
                
        permute(nums, 0)
        return res
                
